client/state/data-layer/wpcom/login-2fa/index.js

```diff
@@ -20,77 +20,83 @@ import {
+			headers: [
+				[ 'Content-Type', 'application/x-www-form-urlencoded' ],
+			],
+			withCredentials: true,
+			body: {
+				user_id: getTwoFactorUserId( store.getState() ),
+				auth_type: authType,
+				two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
+				remember_me: getRememberMe( store.getState() ),
+				two_step_push_token: getTwoFactorPushToken( store.getState() ),
+				client_id: config( 'wpcom_signup_id' ),
+				client_secret: config( 'wpcom_signup_key' ),
+			},
+		},
+		action
+	) );
 };
 
+const receivedTwoFactorPushNotificationApproved = ( { dispatch } ) =>
+	dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
+
 /***
- * Polling the login API for the status of the push notification.
- * The polling will stop on success or when store's polling progress
- * state changes to `false`
+ * Receive error from the two factor push notification status http request
  *
- * @param {Object}   store  Global redux store
+ * @param {Object}	store  Global redux store
+ * @param {Object}	action dispathced action
+ * @param {Object}	error the error object
  */
-const doAppPushPolling = store => {
-	let retryCount = 0;
-	const retry = () => {
-		// if polling was stopped or not in progress - stop
-		if ( ! getTwoFactorPushPollInProgress( store.getState() ) ) {
-			return;
-		}
+const receivedTwoFactorPushNotificationError = ( store, action, next, error ) => {
+	const twoStepNonce = get( error, 'response.body.data.two_step_nonce' );
 
-		retryCount++;
-		setTimeout(
-			() => doAppPushRequest( store ).catch( retry ),
-			( POLL_APP_PUSH_INTERVAL_SECONDS + Math.floor( retryCount / 10 ) ) * 1000 // backoff lineary
-		);
-	};
+	if ( ! twoStepNonce ) {
+		store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_STOP } ); //
+		throw new Error( "Two step nonce wasn't present on the response from polling endpoint, unable to continue" );
+	}
+
+	store.dispatch( {
+		type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
+		nonceType: 'push',
+		twoStepNonce,
+	} );
 
 	if ( getTwoFactorPushPollInProgress( store.getState() ) ) {
-		doAppPushRequest( store ).catch( retry );
+		setTimeout(
+			// eslint-disable-next-line no-use-before-define
+			() => makePushNotificationRequest( store, { type: action.type } ),
+			POLL_APP_PUSH_INTERVAL_SECONDS * 1000
+		);
 	}
 };
 
-const handleTwoFactorPushPoll = ( store ) => {
-	// this is deferred to allow reducer respond to TWO_FACTOR_AUTHENTICATION_PUSH_POLL_START
-	defer( () => doAppPushPolling( store ) );
-};
+const makePushNotificationRequest = dispatchRequest(
+	requestTwoFactorPushNotificationStatus,
+	receivedTwoFactorPushNotificationApproved,
+	receivedTwoFactorPushNotificationError,
+);
 
 export default {
-	[ TWO_FACTOR_AUTHENTICATION_PUSH_POLL_START ]: [ handleTwoFactorPushPoll ],
+	[ TWO_FACTOR_AUTHENTICATION_PUSH_POLL_START ]: [ ( store, action ) => {
+		// We need to store to update for `getTwoFactorPushPollInProgress` selector
+		store.dispatch( local( action ) );
```

Comment:
Related to the comment I left above, I think there's a way to bring all _polling_ knowledge locally into here. That is, polling might be merely responding to the original request, "check on the 2FA push-poll" and if that response indicates that the poll should end then we stop, but if it doesn't, then we re-dispatch the action again after a delay.

When we leave all the polling logic here in this one spot then the app doesn't need to be aware that polling is going on and further it makes it easier for us to imagine replacing the polling with a WebSocket which will just send back the response when it's ready via push instead of poll.